dev/com.ibm.ws.st.liberty.buildplugin.integration/src/com/ibm/ws/st/liberty/buildplugin/integration/internal/AbstractLibertyRuntimeProvider.java

```diff
@@ -0,0 +1,304 @@
+import com.ibm.ws.st.core.internal.WebSphereRuntime;
+import com.ibm.ws.st.core.internal.WebSphereRuntimeLocator;
+import com.ibm.ws.st.core.internal.WebSphereServerInfo;
+import com.ibm.ws.st.core.internal.WebSphereUtil;
+import com.ibm.ws.st.core.internal.config.ConfigurationFile;
+
+/**
+ *
+ */
+@SuppressWarnings("restriction")
+public abstract class AbstractLibertyRuntimeProvider extends LibertyRuntimeProvider {
+
+    /**
+     *
+     */
+    public AbstractLibertyRuntimeProvider() {
+        // Empty
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * If the server.xml file is located ANYWHERE within a supported Liberty Build Plugin project, then 'map' it to the one
+     * that tools knows about. eg. in the build or target folder.
+     */
+
+    @Override
+    public URI getTargetConfigFileLocation(URI uri, WebSphereServerInfo server) {
+        if (uri != null) {
+            IContainer[] findContainersForLocationURI = ResourcesPlugin.getWorkspace().getRoot().findContainersForLocationURI(uri);
+            if (findContainersForLocationURI != null && findContainersForLocationURI.length == 1) { // Should only be the one
+                IContainer container = findContainersForLocationURI[0];
+                IProject project = container.getProject();
+                LibertyBuildPluginConfiguration config = getBuildPluginConfiguration(project);
+                if (config == null) {
+                    return null;
+                }
+
+                WebSphereRuntime webSphereRuntime = getWSRuntime(project, config);
+                // It if is the same runtime, then the server is created (not a Ghost Runtime) so we found the correct server info
+                if (webSphereRuntime != server.getWebSphereRuntime()) {
+                    return null;
+                }
+
+                String serverOutputDirectoryValue = config.getConfigValue(ConfigurationType.serverOutputDirectory);
+                String targetServerConfig = getTargetServerConfig(serverOutputDirectoryValue);
+                IResource findMember = convertConfigValueToIResource(targetServerConfig, project);
+                if (findMember != null && findMember.exists()) {
+                    String uriString = "file:" + findMember.getLocation().toString().replaceAll(" ", "%20");
+                    return URI.create(uriString);
+                }
+                // If not a workspace resource, then return the external URI as is
+                // installDir option where it points to any pre-installed Liberty
+                return URI.create("file:" + targetServerConfig.replaceAll(" ", "%20"));
+            }
+        }
+        return null;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public WebSphereServerInfo getWebSphereServerInfo(URI uri) {
+        try {
+            if (uri != null) {
+                IContainer[] findContainersForLocationURI = ResourcesPlugin.getWorkspace().getRoot().findContainersForLocationURI(uri);
```

Comment:
It is a bit confusing to have a verb in a variable name - can this be called containersForLocationURI or containers?